### ISSUE TITLE
Implement Guzzle6 Bridge

### ISSUE THREAD

**Issue by bshaffer** [google/google-api-php-client#720]
Implement a bridge for Guzzle5 and Guzzle6 as seen in [aws-php-sdk](https://github.com/aws/aws-sdk-php/tree/master/src/Handler) so dependent projects can use either one

**Comment by dwsupplee** [comment]
Hey Brent, since we'll need to be doing something similar in the gcloud library I'd love it if we could sync up on how to move forward with this. :)

**Comment by vinkla** [comment]
:+1:

**Comment by bshaffer** [comment]
Yes, let's!
The way I've envisioned it starts with google/auth. It can follow the psr7 middleware spec (accept request and optional response, return a response). Then we can use a bridge for guzzle5 to implement psr-7. 

Once the auth library implements we can make this library (google/apiclient) do the same. 

This will be a lot of work. 

The other option is to stick with guzzle 5, fork the project, change all the namespaces to prevent conflicts, commit it directly to our projects, and become parias of the community.

**Comment by dwsupplee** [comment]
Well, as tempting as option number 2 sounds - I much prefer a good hard day's work. Sounds like a solid plan!

**Comment by Nyholm** [comment]
@bshaffer and @dwsupplee Im currently active in an other open source project where the following argument came up: 

> I used [Guzzle] 5.2 since google/apiclient requires it.

People are adapting after you and using old version of Guzzle just to be compatible to you. That open source project could not update to Guzzle 6 because of a dependency to `google/apiclient`. This is the reason why you should not be dependent to any implementation. Please consider [PHP-HTTP](http://php-http.readthedocs.org/en/latest/) as a third option.

**Comment by bshaffer** [comment]
Thank you for your feedback @Nyholm. Here are my thoughts on the topic:

First, I agree in order for this library to be useful and stay current with the PHP community, we will need to adopt PSR-7

Second,  **[v1.x](https://github.com/google/google-api-php-client/tree/v1-master) is the *currently supported version*, and does not require guzzle as a dependency**, so if anyone has major conflicts with this library, they should be using that version.

Last, please understand PSR-7 had not been released at the time this library and its dependencies were under development. In addition, PSR-7 asks quite a bit out of libraries in order to adopt it. Decisions were made in PSR-7 to prioritize certain ideals over driving adoption in the community. Some popular PHP libraries have no plans to adopt PSR-7 in their core (I am referring primarily to Symfony and Symfony components). If it was easy to refactor this library for PSR-7, we would have done it by now, but it will require quite a bit of active development to do so. I do not think using `PHP-HTTP` is an exception to this.

Please continue to `+1` this issue, so we can prioritize getting this feature done.

**Comment by Nyholm** [comment]
> First, I agree in order for this library to be useful and stay current with the PHP community, we will need to adopt PSR-7

Great

> Second, v1.x is the currently supported version, and does not require guzzle as a dependency, so if anyone has major conflicts with this library, they should be using that version.

That's good. But let's focus on version 2 for now. 

> Last, please understand PSR-7 had not been released at the time this library and its dependencies were under development. In addition, PSR-7 asks quite a bit out of libraries in order to adopt it. Decisions were made in PSR-7 to prioritize certain ideals over driving adoption in the community. Some popular PHP libraries have no plans to adopt PSR-7 in their core (I am referring primarily to Symfony and Symfony components). If it was easy to refactor this library for PSR-7, we would have done it by now, but it will require quite a bit of active development to do so. I do not think using PHP-HTTP is an exception to this.

You are actually wrong. Symfony has **support** for PSR-7 by a bridge. They (or we) take BC really seriously and it was too late to deprecate the HttpFoundation by the time PSR-7 was voted yes. I have not heard any discussions about not supporting PSR7 in sf4. Yes, sf3 is just about to be released and it is a long time until sf4.0. Anyhow, that is of topic for the discussion...

The fact is that PSR-7 is the standard way of doing HTTP messaging in PHP and `PHP-HTTP` is soon the library that help you to become independent of any transport library like Buzz, Guzzle, Goutte, socket, cURL etc etc. 

**The question is:** Does this widely used library created by very skilled developer form *Google* be among the first to adopt the standards and show the community how thing should be done properly? Or do you want to build more legacy code and push that problem into the future?

Like I've said before: I'm more than happy to help implementing PSR-7 and `PHP-HTTP` if you want to.

**Comment by bshaffer** [comment]
Symfony core has no plans to refactor core components for PSR-7 due to, as you say, BC breaks. This is what I am referring to. 

We are very happy to accept any and all contributions from the community regarding enhancements to this library. There is already an effort [here](https://github.com/google/google-auth-library-php/tree/upgrade-to-guzzle-6) to do so. Realistically, we will need the community to assist here if this is in order to get this done. I was hoping to articulate as much in my previous comment, but apparently I just ruffled some feathers, which was not my intention.

In regards to HTTP-Plug, is it correct that if we adopt PSR-7, HTTP-Plug will be satisfied?

**Comment by Nyholm** [comment]
Great. 

PSR-7 is only interfaces. You need an implementation of that interface. One could use Guzzle's implementation of http://docs.guzzlephp.org/en/latest/psr7.html

Once we have a PSR-7 Request object we could send it with 
`HttpClientDiscovery::find()->sendRequest($request);`

I'll look in to the branch you linked.

**Comment by Nyholm** [comment]
I've sent you a PR to https://github.com/google/google-auth-library-php. There are some TODO's left but al lot of the work is done... 

And of course to make sure the tests runs =)

**Comment by cnelissen** [comment]
Yes please. The 2.0 version is pretty much a non-starter for me because I already require Guzzle 6 due to some other dependencies. I am sure other people will be in the same boat.

**Comment by bshaffer** [comment]
@cnelissen thank you for your feedback. What libraries are you using which have this dependency?

**Comment by gigmaps** [comment]
@bshaffer even https://packagist.org/packages/aws/aws-sdk-php requires Guzzle 5.3.* at minimum
using a composer.json like:
```
{
    "require": {
         "google/apiclient": "^2.0.0@RC",
        "aws/aws-sdk-php": "3.*",
     }
}
```
results in the following composer output:
```
Problem 1
    - Conclusion: remove google/apiclient v2.0.0-RC2
    - Conclusion: don't install google/apiclient v2.0.0-RC2
    - google/apiclient v2.0.0-RC1 requires google/auth v0.3 -> satisfiable by google/auth[v0.3].
    - Installation request for google/apiclient ^2.0.0@RC -> satisfiable by google/apiclient[v2.0.0-RC1, v2.0.0-RC2].
    - Conclusion: don't install aws/aws-sdk-php 3.10.1
    - aws/aws-sdk-php 3.0.0 requires guzzlehttp/guzzle ^5.3 || ^6.0.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.0.1 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.0.2 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.0.3 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.0.4 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.0.5 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.0.6 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.0.7 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.1.0 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.10.0 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.2.0 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.2.1 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.2.2 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.2.3 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.2.4 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.2.5 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.2.6 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.3.0 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.3.1 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.3.2 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.3.3 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.3.4 requires guzzlehttp/guzzle >=5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.3.5 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.3.6 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.3.7 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.3.8 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.4.0 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.4.1 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.5.0 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.6.0 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.7.0 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.8.0 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.8.1 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.8.2 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.9.0 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.9.1 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.9.2 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.9.3 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - aws/aws-sdk-php 3.9.4 requires guzzlehttp/guzzle ~5.3|~6.0.1|~6.1 -> satisfiable by guzzlehttp/guzzle[5.3.0, 6.0.1, 6.0.2, 6.1.0].
    - Can only install one of: guzzlehttp/guzzle[5.3.0, 5.2.0].
    - Can only install one of: guzzlehttp/guzzle[6.0.0, 5.2.0].
    - Can only install one of: guzzlehttp/guzzle[6.0.1, 5.2.0].
    - Can only install one of: guzzlehttp/guzzle[6.0.2, 5.2.0].
    - Can only install one of: guzzlehttp/guzzle[6.1.0, 5.2.0].
    - google/auth v0.3 requires guzzlehttp/guzzle 5.2.* -> satisfiable by guzzlehttp/guzzle[5.2.0].
    - Conclusion: remove guzzlehttp/guzzle 5.2.0|install aws/aws-sdk-php 3.10.1
    - Installation request for aws/aws-sdk-php 3.* -> satisfiable by aws/aws-sdk-php[3.0.0, 3.0.1, 3.0.2, 3.0.3, 3.0.4, 3.0.5, 3.0.6, 3.0.7, 3.1.0, 3.10.0, 3.10.1, 3.2.0, 3.2.1, 3.2.2, 3.2.3, 3.2.4, 3.2.5, 3.2.6, 3.3.0, 3.3.1, 3.3.2, 3.3.3, 3.3.4, 3.3.5, 3.3.6, 3.3.7, 3.3.8, 3.4.0, 3.4.1, 3.5.0, 3.6.0, 3.7.0, 3.8.0, 3.8.1, 3.8.2, 3.9.0, 3.9.1, 3.9.2, 3.9.3, 3.9.4].
```

**Comment by cnelissen** [comment]
Two that I am using right now are influxdb/influxdb-php and mailgun/mailgun-php.

**Comment by Nyholm** [comment]
I have a PR om `mailgun/mailgun-php` to remove the dependency on Guzzle. It is just waiting on `php-http/httplug` to become stable.

**Comment by bshaffer** [comment]
PR #768 represents a branch where guzzle5 and guzzle6 are supported. This is now available for installation [via packagist](https://packagist.org/packages/google/apiclient#dev-add-support-for-guzzle-6) for anyone feeling brave and/or desperate.

**Comment by Behunin** [comment]
You know me so well @bshaffer, I am feeling brave and desperate today :)

**Comment by uxweb** [comment]
HI!, i'm using the dev-add-support-for-guzzle-6 branch.

I'm working with the YouTube Service classes but i'm getting this error:

MediaFileUpload.php line 310:
Call to undefined method GuzzleHttp\Psr7\Request::setHeader()

This is a Laravel 5.1 application and by default it includes guzzle 6.1.1 whichs won't let me install the RC version of this library :(.

Is there a way to use the RC with guzzle 6?

Thanks!

**Comment by bshaffer** [comment]
@uxweb no there isn't, but I've fixed that issue here: 91591b4cddb97d7dd19ab9700cdc964c358b61f2

And you providing feedback like this is incredibly helpful, so please keep it up and we will have an RC soon!

**Comment by bshaffer** [comment]
This has now been merged into master [here](https://github.com/google/google-api-php-client/pull/768) and tagged as [`v2.0.0-RC4`](https://github.com/google/google-api-php-client/releases/tag/v2.0.0-RC4)

You can check it out by running `composer require "google/apiclient:~2.0@RC"`
